File: Data Structures & Algorithms/permutation-string/submission-3.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if (s1.size() > s2.size()) {
            return false;
        }
        
        int i = 0;
        while( i <= s2.size() - s1.size()) {
            if (s1.contains(s2[i])) {
                vector<int> cf(26,0);
                for (auto& x : s1) cf[x-'a']++;
                cout << i << endl;
                for (auto& x: s2.substr(i, s1.size())) {
                    cf[x-'a'] = max(0, cf[x-'a']-1);
                }
                int s = 0;
                for (auto& x: cf) s += x;
                if (s == 0) return true;
                i++;
            } else {
                i++;
            }
        }

        return false;
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.res = 0
        def dfs(node) -> tuple[int, int]:
            if not node:
                return (0, 0)
            nld, nrd = dfs(node.left), dfs(node.right)
            r1 = 1 + max(nld[0], nrd[0])
            r2 = 1 + nld[0] + nrd[0]
            self.res = max(r2, self.res)
            return (r1, r2)
        self.res = max(dfs(root)[1], self.res)
        return self.res - 1
